openscm/scmdataframe/base.py

```diff
@@ -0,0 +1,810 @@
+            )
+        columns[c_name] = col * num_ts
+
+    meta = pd.DataFrame(columns, index=df.columns)
+    return df, meta
+
+
+class ScmDataFrameBase(object):
+    """This base is the class other libraries can subclass
+
+    Having such a subclass avoids a potential circularity where e.g. openscm imports
+    ScmDataFrame as well as Pymagicc, but Pymagicc wants to import ScmDataFrame and
+    hence to try and import ScmDataFrame you have to import ScmDataFrame itself (hence
+    the circularity).
+    """
+
+    DATA_HIERARCHY_SEPARATOR = '|'
+    """str: String used to define different levels in our data hierarchies.
+
+    For example, "Emissions|CO2|Energy|Coal".
+
+    We copy this straight from pyam
+    """
+
+    def __init__(self, data, columns=None, climate_model: str = "unspecified", **kwargs):
+        """Initialize an instance of an ScmDataFrameBase
+
+        Parameters
+        ----------
+        data: IamDataFrame, pd.DataFrame, np.ndarray or string
+            A pd.DataFrame or data file with IAMC-format data columns, or a numpy array of timeseries data if `columns` is specified.
+            If a string is passed, data will be attempted to be read from file.
+
+        columns: dict
+            If None, ScmDataFrameBase will attempt to infer the values from the source.
+            Otherwise, use this dict to write the metadata for each timeseries in data. For each metadata key (e.g. "model", "scenario"), an array of values (one per time series) is expected.
+            Alternatively, providing an array of length 1 applies the same value to all timeseries in data. For example, if you had three
+            timeseries from 'rcp26' for 3 different models 'model', 'model2' and 'model3', the column dict would look like either `col_1` or `col_2`:
+
+            .. code:: python
+
+                >>> col_1 = {
+                    "scenario": ["rcp26"],
+                    "model": ["model1", "model2", "model3"],
+                    "region": ["unspecified"],
+                    "variable": ["unspecified"],
+                    "unit": ["unspecified"]
+                }
+                >>> col_2 = {
+                    "scenario": ["rcp26", "rcp26", "rcp26"],
+                    "model": ["model1", "model2", "model3"],
+                    "region": ["unspecified"],
+                    "variable": ["unspecified"],
+                    "unit": ["unspecified"]
+                }
+                >>> assert pd.testing.assert_frame_equal(
+                    ScmDataFrameBase(d, columns=col_1).meta,
+                    ScmDataFrameBase(d, columns=col_2).meta
+                )
+
+            Metadata for ['model', 'scenario', 'region', 'variable', 'unit'] is required, otherwise a ValueError will be raised.
+
+        kwargs:
+            Additional parameters passed to `pyam.core.read_files` to read nonstandard files
+        """
+        if columns is not None:
+            (_df, _meta) = from_ts(data, **columns)
```

Comment:
```suggestion
            (_df, _meta) = from_ts(data, index=index, **columns)  # requires adding index to the keyword arguments of __init__
```

we need to make a decision about whether we can pass index here or not. I would favour yes, which just means you need to add an index keyword arg which can be used (and is required) if columns is not None